### DISCUSSION TITLE
Cookie issue on server component

### DISCUSSION THREAD

**Discussion by Inayat-Ali** [vercel/next.js#46487] [Help]
I want to set a cookie in middleware and on server component i want to get that cookie and based on that call an api but when i deploy it to vercel it give me "500 | Internal server error" in nextjs version 13.
Not getting cookie.
It is working fine on localhost but after deploying the error occurs

**Comment by icyJoseph** [comment]
Does it work locally if you build, `next build`, and run it with `next start`?

Are you also trying to call an API route that's on the same server? Last but not least, could you please share some code/repository?

**Reply by Inayat-Ali** [reply]
I am setting cookie in middleware:
export function middleware(req: NextRequest){
  const cookie = req.cookies.get("app_id")?.value;
  if(cookie) return NextResponse.next();

  const requestHeaders = new Headers(req.headers);
  const app_id: string | undefined = requestHeaders.get('host')?.split('.')[0];
  requestHeaders.set("app_id", app_id?.toString() || "");

  const response = NextResponse.next({
    request: {
        headers: requestHeaders,
    },
  });
  response.cookies.set("app_id", app_id || "");
    return response;
}
Now i have a page in app directory in which i am using header and cookies from  next to get app_id, it work on local fine but not on vercel production.

Page Code:
async function fetchMethod(category_id: string | number | boolean){
  const url = "https:/rest_api";
  const cookieStore = cookies();  
  const header = headers();

  const app_id_cookie = cookieStore.get("app_id")?.value || header.get("Cookie")?.split("=")[1];
    console.log(app_id_cookie)
  const bodyData = `contact_id=&category=&timestamp=`;
  const response = await fetch(`${url}/?app_id=${app_id_cookie}`,{ method: "POST", body: bodyData, headers:{
    'accept': 'application/json',
    'Content-Type': ' application/x-www-form-urlencoded'
    },
    cache: "force-cache"
  });
 return await response.json();
}

**Reply by Inayat-Ali** [reply]
If I Pass Static Hard code app_id then it work fine on both local and vercel production.

**Reply by Inayat-Ali** [reply]
no response yet

**Reply by icyJoseph** [reply]
Yeah sorry, been super busy. Could you try to setup a reproduction demo? Then we can try to attack the problem, and at least come to a conclusion.

**Reply by Inayat-Ali** [reply]
Yeah sorry for late response, new nextjs 13.2 version was launched and i was trying to implement this with new version using api route handler but did not get solution

**Reply by Inayat-Ali** [reply]
Can you tell me, is it possible to get url part which is before localhost.
like the url is : 1.localhost:3000, i want to call internal api using 1.localhost:3000/api inside fetch but fetch failed error occur.
I am calling it inside my server component in next js v13.2

**Reply by Inayat-Ali** [reply]
my middleware is not working properly in production which i found yet
It is not getting called first in the production when any url browser hit.
